### ISSUE TITLE
softmax function impl for clustering model

### ISSUE THREAD

**Issue by ananis25** [ruiqi-zhong/nlparam#2]
Hi Ruiqi - very interesting idea, I was making my way through the code/paper. 

I am curious about the softmax routine used in clustering. The input is batched along the first dimension, so I expected both the max and sum operation to be along axis 1. They are being calculated globally (max) and along batch axis (sum). What's the reason for it?

https://github.com/ruiqi-zhong/nlparam/blob/41950f0d56758b9d7845ea1afd2dec2adefde597/nlparam/models/cluster_model.py#L16

**Comment by ruiqi-zhong** [comment]
great question. 

the logits shape is N x K, N is the number of datapoints, K is the number of predicates. 

Usually in "classification task", we are classifying which one of the K classes (predicates) an input belongs to.

In our paper, the goal is to define a distribution over the samples x (equation 1). e.g., we are defining K cluster density functions, each parameterized by a predicate and being a distribution over the samples. 

If that resolves your question, you can close this issue; otherwise feel free to let me know if you have further question!

**Comment by ananis25** [comment]
Thank you! I sort of get it, the paper defines a joint distribution over the predicates and the weights (logits) but I gotta look at the theory again. 

I think what confused me a little is this differs from the spherical K-means heuristic.
1. You have N points, and K centroid vectors. 
2. Compute the cosine distance of each point from each centroid. 
3. Assign each point to the closest centroid. 

Running `optimize_w` for the first time, the `phi_denotation` array is indeed the output of 2 above, but step 3 turns out different because of the softmax. 

Thanks again! I'll ask more once I get my head around this.
